// httpunch/Gruntfile.js
'use strict';

var exec = require('child_process').exec,
    fs = require('fs');

module.exports = function (grunt) {
    grunt.initConfig({
        clean: {
            build: ["reports", "coverage"]
        },
        jshint: {
            files: ['Gruntfile.js', 'lib/**/*.js'],
            options: {
                jshintrc: '.jshintrc'
            }
        },
        tape: {
            files: ['test/*.js']
        }
    });

    grunt.registerTask('lint', ['jshint']);
    grunt.loadNpmTasks('grunt-contrib-clean');
    grunt.loadNpmTasks('grunt-contrib-jshint');
    grunt.loadNpmTasks('grunt-tape');

    grunt.registerTask('default', ['test']);

    grunt.registerTask('test', ['jshint', 'tape']);

    grunt.registerTask('coverage', ['clean', 'cover']);

    grunt.registerTask('cover', 'Run coverage from npm script.', function () {
        var done = this.async();

        fs.mkdir('reports', function (error) {
            if (error) {
                grunt.log.warn(error);
            }
            exec('npm run cover', function (error, stdout, stderr) {

                if (error) {
                    grunt.log.error(error);
                    done(error);
                    return;
                }

                stderr && grunt.log.writeln(stderr);
                stdout && grunt.log.writeln(stdout);

                fs.rename('coverage', 'reports/coverage', function (error) {
                    if (error) {
                        grunt.log.error(error);
                        done(error);
                        return;
                    }

                    done(true);
                });
            });
        });
    });
};
