// metrics-ebay/lib/specs/rest.js
'use strict';

module.exports.metrics = {
    statusXXX: {
        operation: 'inc'
    },
    statusXXX_total: {
        ref: 'statusXXX',
        metric: 'cumulative'
    },
    status4XX: {
        operation: 'inc'
    },
    status4XX_total: {
        ref: 'status4XX',
        metric: 'cumulative'
    },
    status5XX: {
        operation: 'inc'
    },
    status5XX_total: {
        ref: 'status5XX',
        metric: 'cumulative'
    },
    status200: {
        operation: 'inc'
    },
    status200_total: {
        ref: 'status200',
        metric: 'cumulative'
    },
    errors: {
        operation: 'inc'
    },
    errors_total: {
        ref: 'errors',
        metric: 'cumulative'
    },
    totalTime: {
        operation: 'add',
        metric: 'mean'
    },
    maxTime: {
        ref: 'totalTime',
        metric: 'max'
    },
    minTime: {
        ref: 'totalTime',
        metric: 'min'
    },
    marshalTime: {
        operation: 'add',
        metric: 'mean'
    },
    unmarshalTime: {
        operation: 'add',
        metric: 'mean'
    },
    timeouts: {
        operation: 'inc'
    },
    timeouts_total: {
        ref: 'timeouts',
        metric: 'cumulative'
    },
    calls: {
        operation: 'inc'
    },
    calls_total: {
        ref: 'calls',
        metric: 'cumulative'
    },
    CallTime99Percentile: {
        ref: 'totalTime',
        metric: 99
    }
};

module.exports.identity = [
    'clientId',
    'service',
    'operation'
];
