// instrument-ebay/lib/service-wrappers/soap/soa-headers.js
'use strict';

var SYS_PREFIX = 'X-EBAY-SOA-';

module.exports = {
    /**
     * Prefix used by all SOA headers.
     */
    SYS_PREFIX: SYS_PREFIX,
    /**
     * Request data format.
     */
    REQUEST_DATA_FORMAT: SYS_PREFIX + 'REQUEST-DATA-FORMAT',
    /**
     * Response data format.
     */
    RESPONSE_DATA_FORMAT: SYS_PREFIX + 'RESPONSE-DATA-FORMAT',
    /**
     * Message protocol (SOAP, etc.).
     */
    MESSAGE_PROTOCOL: SYS_PREFIX + 'MESSAGE-PROTOCOL',
    /**
     * Whether to preserve element ordering (a REST property).
     */
    ELEMENT_ORDERING_PRESERVE: SYS_PREFIX + 'ELEMENT-ORDERING-PRESERVE',
    /**
     * Service operation name.
     */
    SERVICE_OPERATION_NAME: SYS_PREFIX + 'OPERATION-NAME',
    /**
     * Service qname.
     */
    SERVICE_NAME: SYS_PREFIX + 'SERVICE-NAME',
    /**
     * Consumer Name.
     */
    CONSUMER_NAME: SYS_PREFIX + 'CONSUMER-NAME',
    /**
     * Message encoding (character set).
     */
    MESSAGE_ENCODING: SYS_PREFIX + 'MESSAGE-ENCODING',
    /**
     * Indicates that the current response message is an error.
     */
    ERROR_RESPONSE: SYS_PREFIX + 'ERROR-RESPONSE',
    /**
     * List of globalization locale choices (in client request) or single globalization
     * locale selection (in server response).
     */
    LOCALE_LIST: SYS_PREFIX + 'LOCALE-LIST',
    /**
     * Global ID for this request/response.
     */
    GLOBAL_ID: SYS_PREFIX + 'GLOBAL-ID',
    /**
     * Service version in which client (in requests) or server (in responses) is operating.
     */
    VERSION: SYS_PREFIX + 'SERVICE-VERSION',
    /**
     * Full request identifier including chaining path information.
     */
    REQUEST_ID: SYS_PREFIX + 'REQUEST-ID',
    /**
     * Globally unique ID for the request - the uniqueness portion of the request ID.
     */
    REQUEST_GUID: SYS_PREFIX + 'REQUEST-GUID',
    /**
     * Marker - Header sent by client to indicate that service has to send its RLogID in the response to client
     */
    SEND_RLOGID: SYS_PREFIX + 'SEND-RLOGID',
    /**
     * RlogID - a string that is embedded in the webpage to locate CAL logs quickly for debugging purposes.
     */
    RLOGID: SYS_PREFIX + 'RLOGID',
    /**
     * Use case - sent by clients to identify a particular client business use case (e.g. SYI) for
     * reporting purposes
     */
    USECASE_NAME: SYS_PREFIX + 'USECASE-NAME',
    /**
     * ConsumerId - sent by consumers to identify a particular client business for
     * reporting purposes
     */
    CONSUMER_ID: SYS_PREFIX + 'CONSUMER-ID',
    /**
     * PageId - set in the header to identify a particular URL/page for
     * reporting purposes
     */
    PAGE_ID: SYS_PREFIX + 'PAGE-ID',
    /**
     * Tracks server side metrics. At the time of this writing, it only tracks
     * total time taken by the server side code (ser/de-ser/business time etc)
     */
    SERVICE_METRICS: SYS_PREFIX + 'SERVICE-METRICS',
    /**
     * Used in client requests to select a particular response transport to be used
     * by the server.
     */
    RESPONSE_TRANSPORT: SYS_PREFIX + 'RESPONSE-TRANSPORT',
    /**
     * Used in client requests to set the schema validation level to be used
     * by the server.
     */
    REQ_PAYLOAD_VALIDATION_LEVEL: SYS_PREFIX + 'REQ-PAYLOAD-VALIDATION-LEVEL',
    /**
     * Used to abbreviate the naming path in name-value requests; indicates that
     * all names should be considered to begin with the specified prefix.
     */
    NV_IMPLIED_ROOT: SYS_PREFIX + 'NV-IMPLIED-ROOT',
    /**
     * Marks end of headers and beginning of the payload in REST requests
     *
     * This is NOT used as a real transport header
     */
    REST_PAYLOAD: SYS_PREFIX + 'REST-PAYLOAD',
    ALTERNATE_FAULT_STATUS: SYS_PREFIX + 'ALTERNATE-FAULT-STATUS',
    CALLING_POOL : SYS_PREFIX + 'CALLING-POOL',
    /**
     * Indicates whether to buffer the payload before the serilization/deserialization
     * is started.
     */
    NON_STREAMING_MODE:  SYS_PREFIX + 'NON-STREAMING-SERIALIZATION',
    /**
     * Header that stored the messagedigest calculated for request payload
     */
    REQUEST_MD_HASH:  SYS_PREFIX + 'REQUEST-HASH',
    AUTH_ASSERTION: SYS_PREFIX + 'SECURITY-ASSERTION',
    AUTH_COOKIE: SYS_PREFIX + 'SECURITY-COOKIE',
    AUTH_TOKEN: SYS_PREFIX + 'SECURITY-TOKEN',
    AUTH_USERID: SYS_PREFIX + 'SECURITY-USERID',
    AUTH_PASSWORD: SYS_PREFIX + 'SECURITY-PASSWORD',
    AUTH_APPNAME: SYS_PREFIX + 'SECURITY-APPNAME',
    AUTH_ESAMSTOKEN: SYS_PREFIX + 'SECURITY-ESAMSTOKEN',


    AUTH_IAFTOKEN:  SYS_PREFIX + 'SECURITY-IAFTOKEN',

    SOA_HEADER_DEVICE_ID: 'X-EBAY3PP-DEVICE-ID'
};
