// service-instruments-ebay/metrics.js
'use strict';

var MetricsServices;
var logger = require('logging-inc').logger('service-instruments-ebay/metrics');

module.exports = function metrics(pipe) {
    var clientId = pipe.context.clientId;
    var updateMetrics;

    MetricsServices = MetricsServices || require('metrics-ebay/services');
    
    pipe.on('request', function onRequest(request, next) {
        logger.info('metrics handler', clientId);
        var options = request;

        pipe.context.metrics = MetricsServices.getOrCreateMonitor({
            clientId: 'REST',
            service: clientId,
            operation: pipe.context.operation || options.operation ||
                (options.basepath ? options.basepath.replace(/(^\/|\/$)/g, '').replace(/\//g, '.') : options.method) ||
                'unknown'
        });

        // start recording
        updateMetrics = pipe.context.metrics.start();

        next();
    });

    pipe.on('error', function onErr(err, next) {
        // update metrics
        updateMetrics && updateMetrics(err);
        // continue with result to the next response handler
        next();
    });

    pipe.on('response', function onResponse(response, next) {
        // update metrics
        updateMetrics && updateMetrics(null, response);
        // continue with result to the next response handler
        next();
    });

};
