// global-header-ebay/lib/utils.js
'use strict';

var Querystring = require('querystring');

var POOLS = {
        development: 'qa.ebay.com',
        sandbox: 'sandbox.ebay.com',
        qa: 'qa.ebay.com',
        staging: 'qa.ebay.com',
        test: 'qa.ebay.com',
        'pre-production': 'latest.ebay.com',
        production: 'ebay.com',
        xstage: "xstage.ebay.com"
    },
    patterns = {
        script: /<script[^>]*>([\s\S]*?)<\/script>/gim,
        style: /<style[^>]*>([\s\S]*?)<\/style>/gim,
        comment: /<!--(?!<!)([^\[>]*?)-->/gim,
        customAttribute: /\s*_sp=["']?(?:.(?!["']?\s+(?:\S+)=|[>"']))+.["']?/gim
    };

var poolName = POOLS[process.env.NODE_ENV || 'development'];

module.exports = {
    getPoolName: function getPoolName() {
        return poolName;
    },
    getPattern: function getPattern(type) {
        return patterns[type] || new RegExp('a^'); // if pattern not found return a pattern that doesnt match anything
    }
};

module.exports.generateKey = function generateKey(options, prefix) {
    var query = Querystring.stringify(options);
    query = query.replace(/&/g, '__').replace(/=/g, '_');
    return (prefix ? prefix + (query ? '_' : '') : prefix) + query;
};

module.exports.tryParse = function tryParse(text, callback) {
    try {
        return {
            data: JSON.parse(text)
        };
    }
    catch (err) {
        return {
            err: err
        };
    }
};
